src/python/pants/core/target_types.py

```diff
@@ -324,6 +324,26 @@ async def package_archive_target(field_set: ArchiveFieldSet) -> BuiltPackage:
     return BuiltPackage(archive, (BuiltPackageArtifact(output_filename),))
 
 
+# -----------------------------------------------------------------------------------------------
+# `gen_rule` target
+# -----------------------------------------------------------------------------------------------
+
+
+class GenRuleCommandField(StringField):
+    alias = "command"
+    required = True
+
+
+class GenRuleTarget(Target):
+    alias = "gen_rule"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        GenRuleCommandField,
+        Sources,
+    )
+    help = "Execute any external tool for its side effects."
```

Comment:
Should probably specify that since the process is sandboxed, it can only really have external side effects. Also, as with the discussion on https://github.com/pantsbuild/pants/pull/12805#discussion_r705663991, it's worth calling out that its best for any external side effects (poking the network, etc) to be idempotent, since they may be retried or canceled.